src/operations/client_bulk_write/command_builder.ts

```diff
@@ -0,0 +1,302 @@
+import { type Document } from '../../bson';
+import { DocumentSequence } from '../../cmap/commands';
+import { MongoInvalidArgumentError } from '../../error';
+import type {
+  AnyClientBulkWriteModel,
+  ClientBulkWriteOptions,
+  ClientDeleteManyModel,
+  ClientDeleteOneModel,
+  ClientInsertOneModel,
+  ClientReplaceOneModel,
+  ClientUpdateManyModel,
+  ClientUpdateOneModel
+} from './common';
+
+/** @internal */
+export class ClientBulkWriteCommandBuilder {
+  models: AnyClientBulkWriteModel[];
+  options: ClientBulkWriteOptions;
+
+  /**
+   * Create the command builder.
+   * @param models - The client write models.
+   */
+  constructor(models: AnyClientBulkWriteModel[], options: ClientBulkWriteOptions) {
+    this.models = models;
+    this.options = options;
+  }
+
+  /**
+   * Gets the errorsOnly value for the command, which is the inverse of the
+   * user provided verboseResults option. Defaults to true.
+   */
+  get errorsOnly(): boolean {
+    if ('verboseResults' in this.options) {
+      return !this.options.verboseResults;
+    }
+    return true;
+  }
+
+  /**
+   * Build the bulk write commands from the models.
+   */
+  buildCommands(): Document[] {
+    // The base command.
+    const command: Document = {
+      bulkWrite: 1,
+      errorsOnly: this.errorsOnly,
+      ordered: this.options.ordered ?? true
+    };
+    // Add bypassDocumentValidation if it was present in the options.
+    if ('bypassDocumentValidation' in this.options) {
+      command.bypassDocumentValidation = this.options.bypassDocumentValidation;
+    }
+    // Add let if it was present in the options.
+    if ('let' in this.options) {
+      command.let = this.options.let;
+    }
+
+    // Iterate the models to build the ops and nsInfo fields.
+    const operations = [];
+    let currentNamespaceIndex = 0;
+    const namespaces = new Map<string, number>();
+    for (const model of this.models) {
+      const ns = model.namespace;
+      if (namespaces.has(ns)) {
+        operations.push(builderFor(model).buildOperation(namespaces.get(ns) as number));
```

Comment:
This creates a new object for each write model just to build the command, only to immediately discard the builders.  Thoughts on structuring this in a way that allocates fewer objects?

Two ideas:

1. create one of each model builder before the loop, and just re-assign the model and then build it.
2. use functions instead of classes:

```typescript
type Builder = (model: AnyClientBulkWriteModel, index: number) => Document;

const insertOneBuilder: Builder = (model: InsertOneModel, index: number) => { ... };
```

Reply:
I've updated to use functions now instead instantiating the objects.